### ISSUE TITLE
nixos/python3Packages: pytest segfaults across multiple packages on Apple silicon

### ISSUE THREAD

**Issue by sarahec** [NixOS/nixpkgs#423831]
### Nixpkgs version

- Unstable (25.11)

### Describe the bug

Building `python312Packages.gpytorch` fails with multiple similar test failures. The last good build in Hydra is 2025-06-25; something happened between then and now to trigger the failures. (TL;dr it looks like a segfault in `sklearn`:

```sh
       > test/variational/test_nearest_neighbor_variational_strategy.py::TestVNNGPNonInducingData::test_fantasy_call Fatal Python error: Fatal Python error: Segmentation faultSegmentation fault
       >
       >
       >
       > Thread 0x00000001f13adf00 (most recent call first):
       >   File "/nix/store/rfzawqrkzf88q46aa16c77l3wrhy6jgq-python3.12-scikit-learn-1.6.1/lib/python3.12/site-packages/sklearn/metrics/_pairwise_distances_reduction/_dispatcher.py", line 293 in compute
       >   File "/nix/store/rfzawqrkzf88q46aa16c77l3wrhy6jgq-python3.12-scikit-learn-1.6.1/lib/python3.12/site-packages/sklearn/neighbors/_base.py", line 869 in kneighbors
       >   File "/private/tmp/nix-build-python3.12-gpytorch-1.14.drv-0/source/gpytorch/utils/nearest_neighbors.py", line 201 in build_sequential_nn_idx
       >   File "/private/tmp/nix-build-python3.12-gpytorch-1.14.drv-0/source/gpytorch/variational/nearest_neighbor_variational_strategy.py", line 465 in _compute_nn
       >   File "/private/tmp/nix-build-python3.12-gpytorch-1.14.drv-0/source/gpytorch/variational/nearest_neighbor_variational_strategy.py/nix/store/0znljckc2qhpsjs033wjc07xzqiysp3g-pytest-check-hook/nix-support/setup-hook: line 23: 40395 Segmentation fault: 11  /nix/store/q54d9qqlb27g2nf9ja4zsa89gl7b3yvm-python3-3.12.11/bin/python3.12 "${flagsArray[@]}"
```

**UPDATE** The same failure happens in `python3Packages.ezyrb`:
```sh
       > tests/test_reducedordermodel.py ....Fatal Python error: Fatal Python error: Segmentation faultSegmentation fault
       >
       >
       >
       > Thread 0x0000000354e0f000 (most recent call first):
       >   <no Python frame>
       >
       > Thread 0x00000001f13adf00 (most recent call first):
       >   File "/nix/store/rfzawqrkzf88q46aa16c77l3wrhy6jgq-python3.12-scikit-learn-1.6.1/lib/python3.12/site-packages/sklearn/metrics/_pairwise_distances_reduction/_dispatcher.py", line 281 in compute
       >   File "/nix/store/rfzawqrkzf88q46aa16c77l3wrhy6jgq-python3.12-scikit-learn-1.6.1/lib/python3.12/site-packages/sklearn/neighbors/_base.py", line 869 in kneighbors
       >   File "/nix/store/rfzawqrkzf88q46aa16c77l3wrhy6jgq-python3.12-scikit-learn-1.6.1/lib/python3.12/site-packages/sklearn/neighbors/_regression.py", line 243/nix/store/0znljckc2qhpsjs033wjc07xzqiysp3g-pytest-check-hook/nix-support/setup-hook: line 23: 54795 Segmentation fault: 11  /nix/store/q54d9qqlb27g2nf9ja4zsa89gl7b3yvm-python3-3.12.11/bin/python3.12 "${flagsArray[@]}"
       For full logs, run:
         nix log /nix/store/5vikb1dai1h18lz86qv9fr2xly2gwm50-python3.12-ezyrb-1.3.0.post2404.drv
```

### 1 -- Observe the first failure

On a M-series Macintosh:

1. Run `nix-build -A python313Packages.gpytorch --check` (`--check will force a build from source)
2. Observe that the code hangs forever after one of the tests.
3. The source file will be locked for reading, you may need to run `sudo pkill python` to kill the process holding onto it.

### 2 -- Narrow down to the likely root cause

Add the following to `disabledTestPaths`:
```sh
  "test/examples"
  "test/kernels"
```

These both contain hanging tests.

There will be five tests that segfault after this:
```sh
    "test/utils/test_nearest_neighbors.py::TestNNUtil::test_1Dtoycase" # Fatal Python error: Segmentation fault
    "test/utils/test_nearest_neighbors.py::TestNNUtil::test_setup" # Fatal Python error: Segmentation fault
    "test/variational/test_nearest_neighbor_variational_strategy.py::TestVNNGPNonInducingData::test_eval_iteration" # Fatal Python error: Segmentation fault
    "test/variational/test_nearest_neighbor_variational_strategy.py::TestVNNGPNonInducingData::test_eval_iteration" # Fatal Python error: Segmentation fault
    "test/variational/test_lmc_variational_strategy.py::TestDeltaIndexedLMCRobustVGP::test_training_iteration_batch_model" # Fatal Python error: Segmentation fault
```

Each of these five tests fail with a segfault in `sklearn`. Once these are fixed, chances are the tests in `test/examples` and `test/kernels` will no longer hang.

### Expected behaviour

This should build and test without errors or hanging

### Relevant log output

```console
> test/variational/test_nearest_neighbor_variational_strategy.py::TestVNNGPNonInducingData::test_fantasy_call Fatal Python error: Fatal Python error: Segmentation faultSegmentation fault
       >
       >
       >
       > Thread 0x00000001f13adf00 (most recent call first):
       >   File "/nix/store/rfzawqrkzf88q46aa16c77l3wrhy6jgq-python3.12-scikit-learn-1.6.1/lib/python3.12/site-packages/sklearn/metrics/_pairwise_distances_reduction/_dispatcher.py", line 293 in compute
       >   File "/nix/store/rfzawqrkzf88q46aa16c77l3wrhy6jgq-python3.12-scikit-learn-1.6.1/lib/python3.12/site-packages/sklearn/neighbors/_base.py", line 869 in kneighbors
       >   File "/private/tmp/nix-build-python3.12-gpytorch-1.14.drv-0/source/gpytorch/utils/nearest_neighbors.py", line 201 in build_sequential_nn_idx
       >   File "/private/tmp/nix-build-python3.12-gpytorch-1.14.drv-0/source/gpytorch/variational/nearest_neighbor_variational_strategy.py", line 465 in _compute_nn
       >   File "/private/tmp/nix-build-python3.12-gpytorch-1.14.drv-0/source/gpytorch/variational/nearest_neighbor_variational_strategy.py/nix/store/0znljckc2qhpsjs033wjc07xzqiysp3g-pytest-check-hook/nix-support/setup-hook: line 23: 40395 Segmentation fault: 11  /nix/store/q54d9qqlb27g2nf9ja4zsa89gl7b3yvm-python3-3.12.11/bin/python3.12 "${flagsArray[@]}"
```

### Additional context

This is a recent problem -- somewhere between late June and now. The upstream package has been patched by @natsukium in the past to deal with scipy issues. 

Will mark as broken for now on Darwin while we await a fix.

### System metadata

 - system: `"aarch64-darwin"`
 - host os: `Darwin 24.5.0, macOS 15.5`
 - multi-user?: `yes`
 - sandbox: `relaxed`
 - version: `nix-env (Nix) 2.28.4`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/store/i7ykaggappvy548qa4m5aw9msj5cz2vf-source`

### Notify maintainers

@veprbl -- gpytorch maintainer
@natsukium  -- based on relevant work upstream
@psyanticy -- for sklearn
@doronbehar -- for scipy

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)

### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+bug%22+label%3A%226.topic%3A+nixos%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by doronbehar** [comment]
> [@doronbehar](https://github.com/doronbehar) -- for scipy

1. Why am I pinged for this?
2. Also, have you tried discussing this with `gpytorch` maintainers?
3. Are you claiming there's a general issue with pytest and scipy/scikit-learn modules? Have you tried to run these tests outside a Nix build environment? Or even with Nix sandbox disabled?

**Comment by sarahec** [comment]
> > [@doronbehar](https://github.com/doronbehar) -- for scipy
> 
> 1. Why am I pinged for this?
> 2. Also, have you tried discussing this with `gpytorch` maintainers?
> 3. Are you claiming there's a general issue with pytest and scipy/scikit-learn modules? Have you tried to run these tests outside a Nix build environment? Or even with Nix sandbox disabled?

I'm seeing the same error happen in other `scikit-learn`-using packages. I brought you in because 1) you'd likely have something to contribute, and 2) so you could keep an eye out if you see this happening elsewhere in the scipy ecosystem.

**Comment by sarahec** [comment]
Here's another error, this time when building `python3Packages.optuna-dashboard`:
```sh
       > python_tests/preferential/test_samplers.py .Fatal Python error: Segmentation fault
       >
       > Thread 0x00000003502d3000 (most recent call first):
       >   <no Python frame>
       >
       > Thread 0x00000001fc105f00 (most recent call first):
       >   File "/nix/store/brw0knlmj41cxn822agj7mvlsl1m4va7-python3.12-torch-2.7.1/lib/python3.12/site-packages/torch/distributions/multivariate_normal.py", line 187 in __init__
       >   File "/nix/store/qbkvqqdycq76nwx17zjrk7ly3142aai0-python3.12-gpytorch-1.14/lib/python3.12/site-packages/gpytorch/distributions/multivariate_normal.py", line 65/nix/store/0znljckc2qhpsjs033wjc07xzqiysp3g-pytest-check-hook/nix-support/setup-hook: line 23:  2685 Segmentation fault: 11  /nix/store/q54d9qqlb27g2nf9ja4zsa89gl7b3yvm-python3-3.12.11/bin/python3.12 "${flagsArray[@]}"
```

@GaetanLepage Sorry, I need your eyes on this one.

**Comment by sarahec** [comment]
And here comes a similar segfault without pytorch (while building `python3Packages.sqlite-vec` for x86_64-darwin. It's fine during `nix-build` and fails during `nixpkgs-review` (!)

```sh
  File "/nix/store/snwhgxnn0ys84a8bx2gi3f6xrdcx38fs-python3.12-pluggy-1.6.0/lib/python3.12/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/nix/store/snwhgxnn0ys84a8bx2gi3f6xrdcx38fs-python3.12-pluggy-1.6.0/lib/python3.12/site-packages/pluggy/_hooks.py", line 512 in __call__
  File "/nix/store/ymyww3j90qaaxp1c61xw9jx1vwvkpgj3-python3.12-pytest-8.3.5/lib/python3.12/site-packages/_pytest/main.py", line 362 in pytest_runtestloop
  File "/nix/store/snwhgxnn0ys84a8bx2gi3f6xrdcx38fs-python3.12-pluggy-1.6.0/lib/python3.12/site-packages/pluggy/_callers.py", line 121 in _multicall
  File "/nix/store/snwhgxnn0ys84a8bx2gi3f6xrdcx38fs-python3.12-pluggy-1.6.0/lib/python3.12/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/nix/store/snwhgxnn0ys84a8bx2gi3f6xrdcx38fs-python3.12-pluggy-1.6.0/lib/python3.12/site-packages/pluggy/_hooks.py", line 512 in __call__
  File "/nix/store/ymyww3j90qaaxp1c61xw9jx1vwvkpgj3-python3.12-pytest-8.3.5/lib/python3.12/site-packages/_pytest/main.py", line 337 in _main
  File "/nix/store/ymyww3j90qaaxp1c61xw9jx1vwvkpgj3-python3.12-pytest-8.3.5/lib/python3.12/site-packages/_pytest/main.py", line 283 in wrap_session
  File "/nix/store/ymyww3j90qaaxp1c61xw9jx1vwvkpgj3-python3.12-pytest-8.3.5/lib/python3.12/site-packages/_pytest/main.py", line 330 in pytest_cmdline_main
  File "/nix/store/snwhgxnn0ys84a8bx2gi3f6xrdcx38fs-python3.12-pluggy-1.6.0/lib/python3.12/site-packages/pluggy/_callers.py", line 121 in _multicall
  File "/nix/store/snwhgxnn0ys84a8bx2gi3f6xrdcx38fs-python3.12-pluggy-1.6.0/lib/python3.12/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/nix/store/snwhgxnn0ys84a8bx2gi3f6xrdcx38fs-python3.12-pluggy-1.6.0/lib/python3.12/site-packages/pluggy/_hooks.py", line 512 in __call__
  File "/nix/store/ymyww3j90qaaxp1c61xw9jx1vwvkpgj3-python3.12-pytest-8.3.5/lib/python3.12/site-packages/_pytest/config/__init__.py", line 175 in main
  File "/nix/store/ymyww3j90qaaxp1c61xw9jx1vwvkpgj3-python3.12-pytest-8.3.5/lib/python3.12/site-packages/_pytest/config/__init__.py", line 201 in console_main
  File "/nix/store/ymyww3j90qaaxp1c61xw9jx1vwvkpgj3-python3.12-pytest-8.3.5/lib/python3.12/site-packages/pytest/__main__.py", line 9 in 
  File "", line 88 in _run_code
  File "", line 198 in _run_module_as_main
Extension modules: numpy._core._multiarray_umath, numpy.linalg._umath_linalg (total: 2)

/nix/store/597aci5k85simnpr10aijb5lm14gx8hs-pytest-check-hook/nix-support/setup-hook: line 23:  5317 Illegal instruction: 4  /nix/store/q2y1zwszjmsap7r6nbk4dxbcsm47abfx-python3-3.12.11/bin/python3.12 "${flagsArray[@]}"
```

I'm starting to wonder if the common denominator is the `pytestCheckHook` on Darwin. But why these packages?

**Comment by sarahec** [comment]
After consulting with @reckenrode, he pointed out https://github.com/pytorch/pytorch/issues/149201

There's an underlying crash in OpenMP.

$BADWORD

**Comment by sarahec** [comment]
So, per https://github.com/pytorch/pytorch/issues/149201#issuecomment-2776842320 the failure happens when we have two copies of libomp (OpenMP) loaded at the same time. This is happening in `gpytorch`:
```sh
       0x10ff20000 -        0x10ffbffff libomp.dylib (*) <c5c91a03-b2a7-3190-ba30-c18937ab5872> /Volumes/VOLUME/*/libomp.dylib

       0x35b5d8000 -        0x35b677fff libomp.dylib (*) <c5c91a03-b2a7-3190-ba30-c18937ab5872> /Volumes/VOLUME/*/libomp.dylib
```

The derivation adds `openmp` in two places:
`++ lib.optionals stdenv.cc.isClang [ llvmPackages.openmp ]`
`++ lib.optionals rocmSupport [ rocmPackages.llvm.openmp ]`

@reckenrode and I are considering the options to fix this. We'll need to choose one or the other at build time depending on the options.
